enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,486 @@
+
+An important facility of the installation flow is the transition from a CEO deployed etcd to one controlled by pacemaker. The basic transition works as follows:
+1. MCO Extensions are used to ensure that pacemaker, corosync, and resource agents are pre-configured on CoreOS using installation manifests.
+2. Upon detection that the cluster infrastructure is using the DualReplica controlPlaneTopology in the infrastructure config, an in-cluster entity (see open questions regarding whether this should be handled by CEO or an additional operator) will run a command on one of the cluster nodes to initialize pacemaker. The outcome of this is that the resource agent will be started on both nodes.
+3. The aforementioned in-cluster entity will signal CEO to relinquish control of etcd by setting CEO's `managedEtcdKind` to `External`. When this happens, CEO removes the etcd pod from the static pod configs. The resource agents for etcd are running from step 2, and they are configured to wait for etcd pods to be gone so they can restart them using Podman.
+4. The installation proceeds as normal once the pods start.
+If for some reason, the etcd pods cannot be started, then the installation will fail. The installer will need to be able to pull logs from the control-plane nodes to provide context for this failure.
+
+Fencing setup is the last important aspect of the cluster installation. In order for the cluster installation to be successful, fencing should be configured and active before we declare the installation successful. Ideally, the fencing secrets should be made available to the control-plane nodes in the initial pacemaker initialization so that fencing can be configured during step 2. There are a few more critical open questions with this:
+1. Should fencing be made active during the installation or should pacemaker start with it disabled and only enable it after being signaled by the in-cluster entity when the cluster installation is detected as successful?
+2. What mechanism will pacemaker use to get access to the secret linked from the BareMetalHost CRD?
+
+If available via the SaaS offering (not confirmed), ZTP may be evaluated as a future offering. This will need further evaluation to ensure passwords are appropriately handled.
+
+Everything else about cluster creation will be an opaque implementation detail not exposed to the user.
+
+#### Day 2 Procedures
+
+As per a standard 3-node control-plane, OpenShift upgrades and `MachineConfig` changes can not be applied when the cluster is in a degraded state.
+Such operations will only proceed when both peers are online and healthy.
+
+The experience of managing a 2-node control-plane should be largely indistinguishable from that of a 3-node one.
+The primary exception is (re)booting one of the peers while the other is offline and expected to remain so.
+
+As in a 3-node control-plane cluster, starting only one node is not expected to result in a functioning cluster.
```

Comment:
Does this means down time during the upgrade?

> The experience of managing a 2-node control-plane should be largely indistinguishable from that of a 3-node one.

3 nodes OCP  upgrade doesn't involve CP downtime.

Reply:
If implemented properly, no. I will eloborate on this, but there shouldn't be any control plane downtime as along as the nodes gracefully reboot and don't do so simultaneously.